parquet/src/test/java/org/apache/iceberg/parquet/TestVariantReaders.java

```diff
@@ -900,6 +1176,31 @@ private static GenericRecord record(GroupType type, Map<String, Object> fields)
     return record;
   }
 
+  private static <T> List<GenericRecord> elements(Type shreddedType, List<T> elements) {
```

Comment:
What about making this accept `VariantValue` instead of `T`? Then you could check for a particular `PhysicalType` to decide whether the value should be in `value` or `typed_value`:

```java
  for (VariantValue element: elements) {
    Map<String, Object> data = Map.of(element.type() != typeToShred ? "value" : "typed_value", element);
    elementRecords.add(record(elementSchema, data);
  }
```

Reply:
You could also accept a `ValueArray` to transform, rather than a `List` so that you'd be able to generate the data to write based on the expected value.

Reply:
I used `List<VariantValue>` instead of `ValueArray` just to avoid one step to create `ValueArray`. But they should be similar.